File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.rs
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut max = 0;
        let mut low = prices[0];
        let mut high = prices[0];
        for i in prices {
            if i < low {
                low = i;
                high = i;
            } else if i > high {
                high = i;
            }
            max = max.max(high - low);
        }
        max
    }
}

File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.ts
function maxProfit(prices: number[]): number {
    let best = 0;
    let low = prices[prices.length - 1];
    let high = prices[prices.length - 1];
    // Indexing backwards makes it easier
    for (let i = prices.length - 1; i >= 0; i--) {
        if (prices[i] > high) {
            low = prices[i];
            high = prices[i];
        } else if (prices[i] < low) { 
            low = prices[i];
        }
        best = Math.max(best, high - low);
    }
    return best;
};